Format API: add tutorial and document primitives

oandregal opened the pull request:
Adds a new tutorial on how to use the Format API and adds documentation for the primitives used through the examples.

[View docs live](https://github.com/WordPress/gutenberg/blob/add/tutorial-format-api/docs/designers-developers/developers/tutorials/format-api/).

- [x] Skeleton for basic example.
- [x] Document primitives
  - [x] `registerFormatType`
  - [x] `RichTextToolbarButton`
  - [x] `toggleFormat`
- [x] Add English and images for basic example.
- [x] Update TOC and manifest (`npm run docs:build`).

oandregal committed:
Basic tutorial on format API

Adds a new HTML tag to selection.

oandregal committed:
Intro: add English

oandregal committed:
Fix typo

oandregal committed:
Document registerFormatType

oandregal committed:
Add some text

avillegasn reviewed (commented):
This documentation should also include ESNext syntax. IMHO, it's easier to understand.

oandregal committed:
Document RichTextToolbarButton

oandregal committed:
Document toggleFormat

nosolosw commented:
> This documentation should also include ESNext syntax.

Thanks for the suggestion. That's something that I'm definitely considering, time permitting. I do think having both provides the best possible experience for any kind of developer (with or withouth prior knowledge in transpiling, JSX, etc).

oandregal committed:
Finish draft for step 1

oandregal committed:
Improve step 1

- Better narrative flow
- Improve code
- Add image with hover text

oandregal committed:
Step 1: avoid repetitions

oandregal committed:
Add <samp> MDN links

oandregal committed:
Step 2, first draft

oandregal committed:
Use <samp> instead of samp

oandregal committed:
Rename files and add them to the TOC

oandregal committed:
Add generated manifest

oandregal committed:
Better explain what a format is

oandregal committed:
Split docs

oandregal committed:
Update TOC and Manifest

nosolosw commented on docs/designers-developers/developers/tutorials/format-api/2-toolbar-button.md:
```diff
@@ -0,0 +1,37 @@
+
+Let's check that everything is working as expected. Reload the post/page and select a text block. Make sure that the new button was added to the format toolbar, it uses the [editor-code dashicon](https://developer.wordpress.org/resource/dashicons/#editor-code), and the hover text is what you set in the title:
+
+![Toolbar with custom button](/docs/designers-developers/assets/toolbar-with-custom-button.png)
```
Does this URL need to point to GitHub raw CDN or is this fine? cc @gziolo do you happen to know?

gziolo replied:
I have no idea, I would give it a go as is and update if necessary. It takes only 15 minutes to see the handbook updated :)
For what's it is worth, it displays correctly on GitHub.

chrisvanpatten replied:
This should be okay but we have started uploading graphics to the make/design P2, so we may want to consider updating here too.

nosolosw replied:
As per https://wordpress.slack.com/archives/C02RP4WU5/p1547229005149900 it looks like this wouldn't work. Going on to use GitHub CDN while we figure the best strategy going forward.

nosolosw replied:
> It takes only 15 minutes to see the handbook updated :)

It actually takes up to half an hour! See https://wordpress.slack.com/archives/C02QB8GMM/p1547058328181200

gziolo commented:
> This documentation should also include ESNext syntax. IMHO, it's easier to understand.

It all depends on whether someone is familiar with ESNext or not. Let's focus on the tutorial first and then add ESNext as soon it's all ready. The issue with ESNext is that we also would need a tutorial explaining how to setup the project to be able to use it.

chrisvanpatten commented on docs/designers-developers/developers/tutorials/format-api/1-register-format.md:
```diff
@@ -0,0 +1,47 @@
+
+Make that plugin available in your WordPress setup and activate it. Then, load a new page/post. The list of available format types is maintained in the `core/rich-text` store, so you may want to check that your custom format has been made avaliable as well. Run this code in your browser's console:
+
+	wp.data.select( 'core/rich-text' ).getFormatTypes();
```
Can we stick to backtick code fences?

nosolosw replied:
I like that it stands out visually, so the reader knows that's important to pause and check.

chrisvanpatten replied:
Does it actually look visually different in the rendered output though, in the Gutenberg Handbook?

chrisvanpatten commented on docs/designers-developers/developers/tutorials/format-api/1-register-format.md:
```diff
@@ -0,0 +1,47 @@
+
+	wp.data.select( 'core/rich-text' ).getFormatTypes();
+
+It'll return an array containing the format types, including your own.
```
I wonder if it would be useful to end with a paragraph "In the next part of this tutorial, we will do XYZ."

chrisvanpatten commented on docs/designers-developers/developers/tutorials/format-api/2-toolbar-button.md:
```diff
@@ -0,0 +1,37 @@
+
+![Toolbar with custom button](/docs/designers-developers/assets/toolbar-with-custom-button.png)
+
+You may also want to check that upon clicking the button the `toggle format` message is shown in your browser's console.
```
Another spot where "in the next section" would be good to move between parts.

nosolosw replied:
I think this is fine. Though I wouldn't be against adding an "in the next section" paragraph if it doesn't feel repetitive.

chrisvanpatten commented on docs/designers-developers/developers/tutorials/format-api/3-apply-format.md:
```diff
@@ -0,0 +1,43 @@
+
+The expected behavior is that the format will be toggled, meaning that the text selected will be wrapped by a `<samp>` tag if it isn't yet, or the tag will be removed if the selection is already wrapped with the tag. Notice that the button renders a different style depending on whether the selection has the tag or not as well - this is controlled by the `isActive` property of the `RichTextToolbar ...
+
+Probably your browser has already displayed the selection differently once the tag was applied. You may want to use a special style for it, though. So far, you can only target the `samp` element in your CSS to do so. Another option that you could explore is to add a class to the format using the `className` option in [`registerFormatType`](/packages/rich-text/README.md#registerFormatType).
```
```suggestion
Your browser may have already displayed the selection differently once the tag was applied, but you may want to use a special style of your own. So far, you can only target the `samp` element in your CSS to do so. Another option that you could explore is to add a class to the format using the `className` option in [`registerFormatType`](/packages/rich-text/README.md#registerFormatType).
```

chrisvanpatten replied:
It's also super simple but a CSS code block might be nice here. I also wonder if it's even worth building the `className` support into the tutorial, since it'd be good practice to encourage className scoping?

nosolosw replied:
How about updating the text to communicate that clearer?

```suggestion
Your browser may have already displayed the selection differently once the tag was applied, but you may want to use a special style of your own. If the `className` property in [`registerFormatType`](/packages/rich-text/README.md#registerFormatType) is not `null`, the declared class will be added to the new element, so you can target it by class name.
```

nosolosw replied:
Commited a new version.

chrisvanpatten commented on docs/designers-developers/developers/tutorials/format-api/README.md:
```diff
@@ -0,0 +1,13 @@
+# Introduction to the Format API
+
+The purpose of this tutorial is to introduce you to the Format API. The Format API enables developers to add custom buttons to the formatting toolbar and have them apply a _format_ to a text selection.
```
```suggestion
The purpose of this tutorial is to introduce you to the Format API. The Format API makes it possible for developers to add custom buttons to the formatting toolbar and have them apply a _format_ to a text selection.
```

chrisvanpatten reviewed (changes requested):
Overall great; just a few small tweaks for clarity and consistency :)

chrisvanpatten committed:
Update docs/designers-developers/developers/tutorials/format-api/3-apply-format.md

Co-Authored-By: nosolosw <nosolosw@users.noreply.github.com>

chrisvanpatten committed:
Update docs/designers-developers/developers/tutorials/format-api/README.md

Co-Authored-By: nosolosw <nosolosw@users.noreply.github.com>

oandregal committed:
Tweak text to recommend class scoping

oandregal committed:
Use production GitHub URL

The doc parsing mechanisms wouldn't recognize that as an URL.

mkaz committed:
Update README.md

Add Bold as a standard example, which people know and are clear what the formatting toolbar is.

mkaz committed:
Update 1-register-format.md

Minor reword to clarify the section to query the store

mkaz committed:
Update 3-apply-format.md

Typo

mkaz committed:
Update 3-apply-format.md

Minor wording.

mkaz reviewed (approved):
I went through and added a couple of minor wording and fixes.
Looks really good, everything worked well :+1:

nosolosw commented:
@chrisvanpatten your feedback was addressed, so I was wondering whether you had further thoughts or should we merge this in?

oandregal committed:
Fix typo

chrisvanpatten commented:
@nosolosw Sorry just revisiting this (my GH notifications are a mess). I'm assuming it's fine but I'll do another quick scan right now :)

chrisvanpatten commented on docs/designers-developers/developers/tutorials/format-api/1-register-format.md:
```diff
@@ -0,0 +1,49 @@
+# Register a new format
+
+The first thing you're going to do in this tutorial is to register the new format that the plugin intends to apply. WordPress has the [`registerFormatType`](/packages/rich-text/README.md#registerFormatType) primitive to do so.
```
Not a block for merge but I think it might be better to refer to this as a "function" which will be a more familiar term for less technical users.

nosolosw replied:
Done at 59b28818d94a376ebf2240b2ede2739c93d3ed3a

chrisvanpatten replied:
Woot, you rock 🚀

chrisvanpatten commented on docs/designers-developers/developers/tutorials/format-api/1-register-format.md:
```diff
@@ -0,0 +1,49 @@
+
+The first thing you're going to do in this tutorial is to register the new format that the plugin intends to apply. WordPress has the [`registerFormatType`](/packages/rich-text/README.md#registerFormatType) primitive to do so.
+
+Let's prepare a minimal plugin to make this work. Create a new folder and a file named `my-custom-format.php` within it containing the necessary PHP code to register and enqueue the JavaScript assets:
```
Didn't one of @mkaz's tutorials cover setting up a basic plugin? Might be nice to link out to that.

nosolosw replied:
Yeah, the JS tutorial goes into detail about this. It's already referenced a bit before.

chrisvanpatten commented on docs/designers-developers/developers/tutorials/format-api/1-register-format.md:
````diff
@@ -0,0 +1,49 @@
+
+```php
+<?php
+
````
Super minor but in the other tutorials we don't have a newline between the plugin doc block and opening PHP tag. Not a block to merge, just a note for the future :)

nosolosw replied:
04bd2e50d44edc7fee4040eed58517afc230d8b6 removes the space

chrisvanpatten reviewed (approved):
I have a few comments but nothing to block merge. Address them if you want, but otherwise I think this is great. Thanks @nosolosw 🚀

oandregal committed:
Use function for clarity

oandregal committed:
Remove spacing
